docs/development.rst

```diff
@@ -2,9 +2,208 @@
+
+    .. code:: python
+
+        def initialize_model_input(self) -> None:
+            super.initialize_model_input()
+            """
+            TODO Initialize model input by reading input parameters
+            from ``self._parameters`` (see below).
+            """
+
+  - The :func:`~openscm.adapter.Adapter.reset` method resets the model
+    to prepare for a new run. It is called once after each call of
+    :func:`~openscm.adapter.Adapter.run` and to reset the model after
+    several calls to :func:`~openscm.adapter.Adapter.step`.
+
+    .. code:: python
+
+        def reset(self) -> None:
+            super.reset()
+            # TODO Reset the model
+
+  - The :func:`~openscm.adapter.Adapter.run` method runs the model
+    over the full time range (as given by the times set by the
+    previous call to
+    :func:`~openscm.adapter.Adapter.initialize_run_parameters`). You
+    should at least implement this function.
+
+    .. code:: python
+
+        def run(self) -> None:
+            # Do not call ``super.run()`` as we override this method
+            """
+            TODO Run the model and write output parameters to
+            `self._parameters` (see below).
+            """
+
+  - The :func:`~openscm.adapter.Adapter.step` method does a single
+    time step. You can get the current time from
+    :attr:`self._current_time
+    <openscm.adapter.Adapter._current_time>`, which you should
+    increase by the time step length and return its value.
+
+    .. code:: python
+
+        def step(self) -> int:
+            # Do not call ``super.step()`` as we override this method
+            """
+            TODO Do a single time step and write corresponding output
+            parameters to ``self._parameters`` (see below).
+            """
+            self._current_time += YEAR
+            return self._current_time
+
+  - The :func:`~openscm.adapter.Adapter.__del__` method cleans up the
+    adapter. If you need to shutdown the model, please implement this
+    method.
+
+    .. code:: python
+
+        def __del__(self):
+            # TODO Shut down model
+
+
+Reading model and input parameters
+**********************************
+
+Model parameters and input data (referred to as general "parameters"
+in OpenSCM) are **pulled** from the :func:`~openscm.core.ParameterSet`
+provided by the OpenSCM Core. From this, you read the parameters in
+the :ref:`time frame <timeframes>` and :ref:`unit <units>` as needed
+by the specific model; conversion is done interally if possible.
+
+OpenSCM defines a :ref:`set of standard parameters
+<standard-parameters>` to be shared between different SCMs. You should
+use these where appropriate.
+
+Refer to the :ref:`low-level-interface` documentation on how to do so.
+
+
+Writing output parameters
+*************************
+
+In the :func:`~openscm.adapter.Adapter.run` and
+:func:`~openscm.adapter.Adapter.step` function you should directly
+write the relevant output parameters. Refer to the
```

Comment:
The output parameters should be written to a ParameterSet. Should we use the same as for input or have a separate one?

Reply:
@znicholls @rgieseke ?

Reply:
> Should we use the same as for input or have a separate one?

Separate I think to avoid overwriting (sometimes parameters get changed/used internally in cryptic ways by the model *cough*MAGICC and HECTOR*cough*) and to make debugging easier. See comment above

Reply:
all right ;)